src/Optimizely/Event/Builder/EventBuilder.php

```diff
@@ -60,147 +58,174 @@ class EventBuilder
-                        'name' => RESERVED_ATTRIBUTE_KEY_BUCKETING_ID_EVENT_PARAM_KEY,
-                        'type' => 'custom',
-                        'value' => $attributeValue,
-                        'shouldIndex' => true
-                    ]);
+                   $feature = [
+                        'entity_id' => RESERVED_ATTRIBUTE_KEY_BUCKETING_ID,
+                        'key' => RESERVED_ATTRIBUTE_KEY_BUCKETING_ID_EVENT_PARAM_KEY,
+                        'type' => CUSTOM_ATTRIBUTE_FEATURE_TYPE,
+                        'value' => $attributeValue
+                        ];
+
                 } else {
                     $attributeEntity = $config->getAttribute($attributeKey);
                     if (!is_null($attributeEntity->getKey())) {
-                        array_push($this->_eventParams[USER_FEATURES], [
-                            'id' => $attributeEntity->getId(),
-                            'name' => $attributeKey,
-                            'type' => 'custom',
+                       $feature = [
+                            'entity_id' => $attributeEntity->getId(),
+                            'key' => $attributeKey,
+                            'type' => CUSTOM_ATTRIBUTE_FEATURE_TYPE,
                             'value' => $attributeValue,
-                            'shouldIndex' => true
-                        ]);
+                        ];
                     }
                 }
             }
+
+            if(!empty($feature))
+                $commonParams[VISITORS][0]['attributes'][] = $feature;
         }
+
+        return $commonParams;
     }
 
     /**
-     * Helper function to set parameters specific to impression event.
+     * Helper function to get parameters specific to impression event.
      *
      * @param $experiment Experiment Experiment being activated.
      * @param $variationId string
      */
-    private function setImpressionParams(Experiment $experiment, $variationId)
+    private function getImpressionParams(Experiment $experiment, $variationId)
     {
-        $this->_eventParams[LAYER_ID] = $experiment->getLayerId();
-        $this->_eventParams[DECISION] = [
-            EXPERIMENT_ID => $experiment->getId(),
-            VARIATION_ID => $variationId,
-            IS_LAYER_HOLDBACK => false
+        try {
+            // Generate a version 4 (random) UUID object
+            $uuid4 = Uuid::uuid4();
+            //echo $uuid4->toString() . "\n"; // i.e. 25769c6c-d34d-4bfe-ba98-e0ee856f3e7a
+        } catch (UnsatisfiedDependencyException $e) {
+            // Some dependency was not met. Either the method cannot be called on a
+            // 32-bit system, or it can, but it relies on Moontoast\Math to be present.
+            echo 'Caught exception: ' . $e->getMessage() . "\n";
+        }
+
+        $impressionParams = [
+            'decisions' => [
+                [
+                    'campaign_id' => $experiment->getLayerId(),
```

Comment:
It's better to replace these string literals with the constants defined in Params.php  E.g: CAMPAIGN_ID. Do the same for all the string literals.